packages/arb-ts/src/lib/assetBridger/tokenBridger.ts

```diff
@@ -0,0 +1,719 @@
+  private async withdrawTxOrGas<T extends boolean>(
+    params: TokenWithdrawParams,
+    estimate: T
+  ): Promise<BigNumber | ethers.ContractTransaction> {
+    if (!SignerProviderUtils.signerHasProvider(params.l2Signer)) {
+      throw new MissingProviderArbTsError('l2Signer')
+    }
+
+    const to = params.destinationAddress || (await params.l2Signer.getAddress())
+
+    const l2GatewayRouter = L2GatewayRouter__factory.connect(
+      this.l2Network.tokenBridge.l2GatewayRouter,
+      params.l2Signer
+    )
+
+    return (estimate ? l2GatewayRouter.estimateGas : l2GatewayRouter.functions)[
+      'outboundTransfer(address,address,uint256,bytes)'
+    ](params.erc20l1Address, to, params.amount, '0x', params.overrides || {})
+  }
+
+  /**
+   * Estimate gas for withdrawing tokens from L2 to L1
+   * @param params
+   * @returns
+   */
+  public async withdrawEstimateGas(
+    params: TokenWithdrawParams
+  ): Promise<BigNumber> {
+    return this.withdrawTxOrGas(params, true)
+  }
+
+  /**
+   * Withdraw tokens from L2 to L1
+   * @param params
+   * @returns
+   */
+  public async withdraw(
+    params: TokenWithdrawParams
+  ): Promise<L2ContractTransaction> {
+    const tx = await this.withdrawTxOrGas(params, false)
+    return L2TransactionReceipt.swivelWait(tx)
+  }
+
+  /**
+   * Fetch a batch of token balances
+   * @param l1OrL2Provider
+   * @param userAddr
+   * @param tokenAddrs
+   * @returns
+   */
+  public async getTokenBalanceBatch(
+    l1OrL2Provider: Provider,
+    userAddr: string,
+    tokenAddrs: Array<string>
+  ): Promise<(BigNumber | undefined)[]> {
+    const network = await l1OrL2Provider.getNetwork()
+    let isL1: boolean
+    if (network.chainId === Number.parseInt(this.l2Network.chainID)) {
+      isL1 = false
+    } else if (
+      network.chainId === Number.parseInt(this.l2Network.partnerChainID)
+    ) {
+      isL1 = true
+    } else {
+      throw new ArbTsError(
+        `Unexpected chain id. Provider chain id: ${network.chainId} matches neither l1Network: ${this.l2Network.partnerChainID} not l2Network: ${this.l2Network.chainID}.`
+      )
+    }
+
+    const multiCaller = new MultiCaller(
+      l1OrL2Provider,
+      isL1
+        ? this.l2Network.tokenBridge.l1MultiCall
+        : this.l2Network.tokenBridge.l2Multicall
+    )
+
+    const callInputs = tokenAddrs
+      .map(t =>
+        isL1
+          ? this.getL1TokenContract(l1OrL2Provider, t)
+          : this.getL2TokenContract(l1OrL2Provider, t)
+      )
+      .map(t => ({
+        targetAddr: t.address,
+        encoder: () => t.interface.encodeFunctionData('balanceOf', [userAddr]),
+        decoder: (returnData: string) =>
+          t.interface.decodeFunctionResult(
+            'balanceOf',
+            returnData
+          )[0] as Awaited<ReturnType<typeof t['balanceOf']>>,
+      }))
+
+    return await multiCaller.call(callInputs)
+  }
+}
+
+/**
+ * A token-gateway pair
+ */
+interface TokenGateway {
```

Comment:
This confused me below - maybe name it TokenToGatewayPairing or something?

Reply:
woops, yep. Renamed to `TokenAndGateway`